src/main/java/picard/sam/AbstractAlignmentMerger.java

```diff
@@ -751,54 +765,103 @@ private void transferAlignmentInfoToPairedRead(final SAMRecord firstUnaligned, f
+                        clipRead(neg, negClipFrom, useHardClipping);
                     }
                 }
             }
         }
     }
 
-    /** Returns the number of soft-clipped bases until a non-soft-clipping element is encountered. */
-    private static int lengthOfSoftClipping(Iterator<CigarElement> iterator) {
-        int clipped = 0;
-        while (iterator.hasNext()) {
-            final CigarElement elem = iterator.next();
-            if (elem.getOperator() != CigarOperator.SOFT_CLIP && elem.getOperator() != CigarOperator.HARD_CLIP) {
-                break;
-            }
-            if (elem.getOperator() == CigarOperator.SOFT_CLIP) {
-                clipped = elem.getLength();
+    private static int getReadPositionAtReferencePositionIgnoreSoftClips(final SAMRecord rec, final int pos) {
+        final int readPosition;
+        final Cigar oldCigar = rec.getCigar();
+        final int oldStart = rec.getAlignmentStart();
+        final Cigar newCigar = new Cigar();
+        List<CigarElement> cigarElements = new ArrayList<>(oldCigar.getCigarElements()); //need to be modifiable
+        int posShift = 0;
+        boolean foundNonClip = false;
+        for (final CigarElement cigarElement : cigarElements) {
+            final CigarOperator op = cigarElement.getOperator();
+
+            if (op == CigarOperator.SOFT_CLIP) {
+                newCigar.add(new CigarElement(cigarElement.getLength(), CigarOperator.MATCH_OR_MISMATCH));
+                if (!foundNonClip) {
+                    posShift += cigarElement.getLength();
+                }
+            } else {
+                if (!op.isClipping()) {
+                    foundNonClip = true;
+                }
+                newCigar.add(new CigarElement(cigarElement.getLength(), op));
             }
         }
 
-        return clipped;
+        rec.setAlignmentStart(rec.getAlignmentStart() - posShift);
+        rec.setCigar(newCigar);
+        readPosition = SAMRecord.getReadPositionAtReferencePosition(rec, pos, false);
+        rec.setCigar(oldCigar);
+        rec.setAlignmentStart(oldStart);
+        return readPosition;
+    }
+
+    private static void clipRead(final SAMRecord rec, final int clipFrom, final boolean useHardClipping) {
+        // If we are using hard clips, add bases and qualities to SAM tag.
+        if (useHardClipping) {
+            moveClippedBasesToTag(rec, clipFrom);
+        }
+
+        // Actually clip the read
+        CigarUtil.clip3PrimeEndOfRead(rec, clipFrom, useHardClipping ? CigarOperator.HARD_CLIP : CigarOperator.SOFT_CLIP);
+    }
+
+    private static void moveClippedBasesToTag(final SAMRecord rec, final int clipFrom) {
+        if (rec.getAttribute(HARD_CLIPPED_BASES_TAG) != null || rec.getAttribute(HARD_CLIPPED_BASE_QUALITIES_TAG) != null) {
+            throw new PicardException("Record already contains tags for restoring hard clipped bases.  This operation will permanently erase information if it proceeds.");
```

Comment:
make sure you include the name of the offending record, and the names of the tags that you need...otherwise, debugging this might be tough.